edit: allow interactive editing during the importer

diego-plan9 opened the pull request:
As a follow up to #396, an attempt to allow the user to launch the edit plugin during an interactive import session.

At the moment the code is still not complete and rather unpolished and full of TODOs and other comments - but hopefully it will serve as a way to discussing some issues that are not as straight-forward as I'd hoped:
* during an interactive import, the `Items` do not have a valid `id` field, as they are not on the database - and the `edit` plugin relied on it for reconciling back the user changes and other internal functionality. I have resorted to using the `path` field instead of `id` when the plugin is invoked on interactive mode, as it seems to satisfy the requirements for a "reference/mapping" field (unique for each item, non-editable by the user). This is done mainly via the introduction of the `EditPlugin.mapping_field` (and using the original items for showing changes during `edit_objects`).
* I have tried to allow the user two main choices - edit based the tags of the files to be imported directly (currently `eDit` -> `importer_edit`) and edit based on a candidate (currently `edit Candidate` -> `importer_edit_candidate`), as they seemed to be two rather common use cases (cleanup files, and fix MusicBrainz problems).
  * The main challenge is the first one, as I could not really find a way to implement it while keeping all the code within the plugin (hence the dirty modification to `importer.py`): editing the Item tags is similar to ASIS ... but not quite. If marked as an ASIS, I could not find an importer code path where the `Items` are really written(), making the changes to be stored correctly on the database but not on the files. If marked as an APPLY, the candidate takes preference and hence the user edits are discarded in favor of the first candidate.
  * For the second one, the implementation currently simply forces an `task.apply_metadata()` before invoking the edits so the `Items` are modified temporarily, and then proceed as in the first case. It is probably more correct (and doable!) to just set a flag on the plugin during `importer_edit_candidate`, let the importer proceed, and actually invoke the editor later (on `import_task_apply` or a new pipeline stage), which would also get rid of the current extra prompt for selecting a candidate.

I'm wondering if you could point me in the right direction for solving the "edit item tags" properly? I have played around with delaying the invoking of the editor until another event for that case as well, but unfortunately could not find an event that is launched at the right moment reliably (using ASIS, and taking into account the different write/copy/link options). Also, the pipeline stage (which could be a good candidate) does not seem to be reachable for an ASIS choice.

diego-plan9 commented:
I forgot to mention that another approach could be to make the edit plugin work on the candidates directly instead of the items (creating a "fake" candidate populated based on the items if needed) - this would play nicer with the importer, be self-contained, and have some added niceness (such as displaying the changes in the same way as the regular importer would), although the main drawback is that some information (ie. file tags that are not mapped onto the candidates) *might* be lost in the process.

sampsyo commented:
> The main challenge is the first one, as I could not really find a way to implement it while keeping all the code within the plugin

Great point. Exactly as you say, the new action here is somewhere halfway between the ASIS action (use the tags as they currently are on the Items) and the APPLY action (which invokes the bits that move files, write tags, and save to the database). Maybe we need a new action type for exactly situation: don't apply any match, but do record new metadata.

(You can imagine other plugins using a similar mode when they have some way of automatically applying metadata independent of match candidates.)

I also like your "fake candidate" idea, which seems quite elegant to me—we could try it and see if the un-applied fields issue becomes a problem.

> and actually invoke the editor later (on import_task_apply or a new pipeline stage),

Sadly, all user interaction has to happen in the `choose_match` stage. Otherwise, there will be multiple threads trying to talk to the user at once.

sampsyo commented:
And I forgot to write: thank you for looking into this! :sparkles: This is a thorny issue, and an important feature, so you're brave to jump right in.

diego-plan9 commented:
Thanks for the thoughtful and useful comments, as usual!

The latest commits intend to add a new `RETAG` (temptative name) action, which basically signals the importer not to apply metadata from the candidates and reach `manipulate_files()` properly, writing the changes to disk if needed. Other than that, it isolates the dealing with the "reference" field into `_set_reference_field`, tidying up the code a bit. Could you please take a look and let me know if this goes more or less along the solution you suggested?

I have also made an attempt at working on the "edit candidates instead of items" approach, but turned out to involve quite a refactoring of the plugin as it currently depends on a bunch of properties of `DBModel` (`o.keys()`, `o[key]`, `o.key`, `o.read()`, formatting, `ui_show_changes()`, etc). Nothing struck me as too difficult or unsolvable, though (the way I see it, it would basically consist of having a couple of different adapter/strategies for each kind of object to be edited, moving some of the existing functions there and maybe splitting some existing functions) - I mainly abandoned it temporarily as I could not invest enough time on it yet, and did want to try if the "new action" approach would suffice first. 

One issue that would ideally need to be discussed regardless of the approach would be how to include the "edit album" (as oposed to edit items) choice without cluttering the interface - the simplest way I can think of is just doing two "editing rounds" always (one for the album fields, one for the item fields), with the rationale that users willing to edit something will already expect some manual input and wouldn't mind having to do a couple more of keypresses if they are not interested in editing one of the entities (album or items). Another option could be having a new prompt whenever the `edit` choice is selected, something like:
```
[A]pply, More candidates, Skip, Use as-is, as Tracks, Group albums,
Enter search, enter Id, aBort, eDit? d

e[D]it items, edit Album, [1-5]...?```
which might also make it more consistent by always having just `"eDit"` on the main prompt.

Acknowledged TODOs: cleanup comments, revise docstrings (plugin and importer functions that use the new action), add tests, take into account the editing of singletons, add plugin documentation.

sampsyo commented on beets/importer.py:
```diff
@@ -672,7 +673,7 @@ def manipulate_files(self, move=False, copy=False, write=False,
                     item.move(copy, link)
 
-            if write and self.apply:
+            if write and (self.apply or self.choice_flag == action.RETAG):
```
Or perhaps `self.apply` should be set in the RETAG case? (I'm not certain; it's just worth considering.)

diego-plan9 replied:
I'm a bit uncertain about what you mean: modifying the `task.apply()` method so it returns True for both `self.choice_flag in (action.APPLY, action.RETAG)`? I wouldn't oppose to that or have a strong opinion (didn't go through the implications yet), as I merely opted for the most explicit and simple way possible of adding the new action.

sampsyo replied:
Yep, that's exactly what I meant. I also don't have a strong opinion; I was just noticing that we were checking `flag in (APPLY, RETAG)` quite often, so it might make sense to use a shortcut.

diego-plan9 replied:
Thanks for the clarification, and yes, indeed makes sense. However, it seems it would require a small modification to `importer.apply_choice()` in order to avoid incorrectly applying the metadata changes if it is a `RETAG` (something similar to):
```diff
@@ -1353,7 +1353,7 @@ def apply_choice(session, task):
         return
 
     # Change metadata.
-    if task.apply:
+    if task.apply and task.choice_flag == action.APPLY:
         task.apply_metadata()
         plugins.send('import_task_apply', session=session, task=task)
```
and probably a similar tweak in regards to the `align_album_level_fields` issue at https://github.com/beetbox/beets/pull/1846#discussion-diff-51828422.

Both the existing approach and your suggested one sound like a reasonable trade-off to me!

diego-plan9 replied:
After checking this issue again, I have decided to keep the explicit comparison with `RETAG` at the moment, mainly because the behaviour of `RETAG` does not always fall under the `APPLY` choice:

| function                   | follows |
|----------------------------|:-------:|
| chosen_ident()             |   ASIS  |
| imported_items()           |   ASIS  |
| align_album_level_fields() |  APPLY  |
| manipulate_files()         |  ~APPLY |

It seemed to me that merging both on `task.apply()` would end up resulting in more confusion and a harder to follow flow! Again, it's rather subjective, so please let me know if you prefer the other approach.

sampsyo replied:
Great; thanks for looking into it. This is a very reasonable conclusion.

sampsyo commented:
Fantastic! This is definitely the right direction. Thanks for exploring all the possibilities—I'm convinced this is the way to go.

> One issue that would ideally need to be discussed regardless of the approach would be how to include the "edit album" (as oposed to edit items) choice without cluttering the interface

That is indeed tricky. Two unsatisfying hacks come to mind:

* Take the album-level data from the first item.
* Include a separate section for album fields in the temporary text file—basically your idea, but two sections in one file instead of two separate files.

Very tricky! There's no obvious answer here.

diego-plan9 commented:
> Fantastic! This is definitely the right direction. Thanks for exploring all the possibilities—I'm convinced this is the way to go.

Great! I'd still be up for giving another try at the "edit candidates" approach if needed, if in the end we are not happy with the results.

> That is indeed tricky. Two unsatisfying hacks come to mind:

> * Take the album-level data from the first item.
> * Include a separate section for album fields in the temporary text file—basically your idea, but two sections in one file instead of two separate files.

Not ideal indeed, but I think the second option is a reasonable compromise and does not look too hard to pull off! I'll work in that direction, at least until a more creative solution appears.

Incidentally, I'm wondering if something should (and can) be done for allowing the plugin to receive command line arguments when not run as a SubCommand: in particular, I see the potential for using the `--all` or `--field` arguments when running the plugin from the importer (user can always resort to editing the config.yaml file, but might be a bit inconvenient, specially for the `--all` flag).

sampsyo commented:
> I'm wondering if something should (and can) be done for allowing the plugin to receive command line arguments when not run as a SubCommand

It's probably best to just make those part of the Python interface to the plugin code—I don't see a clear way to re-use the argument-parsing logic without making things too messy. For this case, I suggest we don't worry about `--all` and `--field` for now; changing the config file seems like an OK workaround for the first iteration.

diego-plan9 commented:
The latest commits provide an implementation for:
> Include a separate section for album fields in the temporary text file—basically your idea, but two sections in one file instead of two separate files.

It basically appends a "fake" Album object when running during the importer to the list of objects to be considered on the yaml, propagating the changes to the items if the editing was successful. The implementation relies on a `FakeAlbum` object that inherits from `lib.Album` in order not to alter the edit plugin flow too much - it is still not covered by the tests, and does some not so elegant things such as overriding the getter for the `path` field.

There are also some further modifications to the importer, in order to allow `SingletonImporterTask` to use the new `RETAG` action properly, and a slight tweak on `ImportTask.align_album_level_fields` which is a bit fragile: is there a better way to find out if, at that stage, we come from a `RETAG-ASIS` or a `RETAG-APPLY`?

Finally, added some tests for executing the plugin during an interactive importer run, in order to provide some confidence for further modifications (that, as mentioned, do **not** include probably-needed tests for the "editing album and item fields at the same time" feature), and also cleaned up and reorganized the previous test case a bit.

I'm running into a slight problem with the tests: it seems that the plugin is not really unloaded between executions despite the `unload_plugins()` call, with all the instances from previous tests receiving the `before_choose_candidate` events (and the garbage collector showing that even after the `unload_plugins`, the instances do have some references lying around). I'm wondering if there are some extra steps that need to be performed to really unload the plugin?

sampsyo commented:
Cool! I'll do a full review soon.

About that testing issue: I don't know of a better way currently. Maybe we need to fix `unload_plugins` to do this by default?

diego-plan9 commented:
Thanks, looking forward to the review!

> About that testing issue: I don't know of a better way currently. Maybe we need to fix unload_plugins to do this by default?

It seems doable (iterate over `beets.plugins._classes` and set `listeners` to None for each item?) and harmless at a first glance, so maybe it is a good idea indeed! A bit off-topic, but has there ever been a demand for a `BeetsPlugin.unregister_listener()` function?

sampsyo commented:
> A bit off-topic, but has there ever been a demand for a BeetsPlugin.unregister_listener() function?

Hmm, not except for during testing. I would hope that plugins wouldn't need to dynamically add and remove listeners all the time for correctness—that seems like a bit of a complexity nightmare.

sampsyo commented on beets/importer.py:
```diff
@@ -620,8 +621,12 @@ def align_album_level_fields(self):
+        # Determine where to gather the info from for the RETAG action.
+        retag_asis = (self.choice_flag == action.RETAG and
+                      not self.items[0].artist and
+                      not self.items[0].mb_artistid)
```
I don't quite follow why these fields are the right ones to check for this... is there a particular reason why you chose `artist` and `mb_artistid`?

diego-plan9 replied:
Sorry: `artist` should have been `albumartist`, in a lazy attempt to just use the same fields that the `elif` is using.

The ultimate reason is trying to find out if we get here by doing a `RETAG-based-on-asis` or 
`RETAG-based-on-candidate`, as there is no `self.match` in either case. I'm not sure if this can be achieved without further "polluting" the class (by adding a flag variable of sorts) or similar (the `match` doesn't get returned from `commands.choose_candidate()` either, just `action.RETAG`). Even if it is not bulletproof, it seemed like a good compromise to check the `mb_artistid` existence in particular as a decent indicator that we are coming from `RETAG-based-on-candidate`.

sampsyo replied:
Ah, thanks for the explanation---I understand now.

Here's a crazy proposal: what if we never apply the "as-is fixup" logic for the RETAG case? The philosophy would be: if you're willing to edit the data manually, then you can take matters into your own hands.

sampsyo commented on beetsplug/edit.py:
```diff
@@ -151,6 +154,23 @@ def __init__(self):
+        elif field == 'path':
+            self.reference_field = 'path'
+            self.ref_field_value = lambda o: util.displayable_path(o.path)
+            self.obj_from_ref = lambda d: util.displayable_path(d['path'])
```
Hmm... It's unfortunate that this needs to update fields on the plugin object. Is there any way we can control this instead by just passing a parameter to the relevant functions? That should make the logical flow easier to read (and, hopefully, it should make for easier testing).

sampsyo commented on beetsplug/edit.py:
```diff
@@ -216,20 +239,26 @@ def edit(self, album, objs, fields):
             self.save_changes(objs)
 
-    def edit_objects(self, objs, fields):
+    def edit_objects(self, objs, item_fields, album_fields):
```
While I'm not *thrilled* about the need to introduce two field sets here, I can see that it might be necessary. (Previously, it was nice that this method just worked on Model objects abstractly; it didn't know or care about Item and Album specifically.)

sampsyo commented on beetsplug/edit.py:
```diff
@@ -262,10 +291,20 @@ def edit_objects(self, objs, fields):
                 # Show the changes.
+                # If the objects are not on the DB yet, we need a copy of their
+                # original state for show_model_changes.
+                if all(not obj.id for obj in objs):
```
Since we need to do this for the not-in-DB case, maybe we should just do it unconditionally? Having this `if` here doesn't seem to buy us much...

diego-plan9 replied:
Agreed!

diego-plan9 replied:
On a closer look, seems like a `deepcopy` of an object that **is** on the library might not be a good idea:
`TypeError: object.__new__(thread.lock) is not safe, use thread.lock.__new__()`

sampsyo replied:
Ah indeed, that does seem problematic. Maybe we should just be keeping a dictionary of data (i.e, `dict(obj)` instead of a full copy of the object?

diego-plan9 replied:
In the end, using a `dict(obj)` didn't work either - the main issue is that we need to use something that is understood by `ui.show_model_changes`, and I'm not really sure we can use anything other than a DBModel without modifying that function (or providing a similar alternative that works on plain dicts, etc).

In any case, I have tidied up a bit the code by always generating `objs_old` (with a `deepcopy` if the object is not in DB, `None` otherwise) and removing the `if`s by iterating both lists at the same time - hope it makes sense.

sampsyo replied:
OK! Once again, thank you for investigating and discarding my half-baked ideas. :smiley: This solution looks very straightforward.

sampsyo commented on beetsplug/edit.py:
```diff
@@ -313,8 +352,9 @@ def apply_data(self, objs, old_data, new_data):
-            id = int(old_dict['id'])
-            apply(obj_by_id[id], new_dict)
+            # Reconcile back the user edits, using the reference_field.
+            val = self.obj_from_ref(old_dict)
```
Continuing the theme, I'm also in general somewhat saddened by the need for all this extra complexity to identify objects by an appropriate field.

Here's one other crazy idea: give the objects a *temporary id* for the purpose of this workflow. Then, after the process is done, remove it before continuing with the import process.

diego-plan9 replied:
> Here's one other crazy idea: give the objects a temporary id for the purpose of this workflow. Then, after the process is done, remove it before continuing with the import process.

That would certainly help lower the amount of complexity introduced on this pull request (taking care of https://github.com/beetbox/beets/pull/1846#discussion-diff-51828719 as well), and something that crossed my mind but felt like dangerous territory.

Are there any special considerations needed to be taken into account when assigning the temporary ids (ie. total_item_count + i)?

sampsyo replied:
Hmm. As long as the IDs never actually get added to the database, I think they can be anything unique---so even just "i" should be a fine way to assign them. Then the IDs can be removed again before they objects get added to the database (at which point they'll be assigned real IDs).

sampsyo commented:
Seeing the complexity it's caused, I wonder if we shouldn't reconsider the feature that lets the user edit album-level fields as well as item-level ones. I know that it's useful (and I originally suggested it!), but this may be a case where the benefit is outweighed by the extra complexity in the code (FakeAlbum and all that).

Maybe we should revisit this as part of a later change—maybe a more general one, since the same kind of thing would be useful during an explicit `beet edit -a` command also?

Beyond that, this is looking like it's on the right track!

diego-plan9 commented:
> Seeing the complexity it's caused, I wonder if we shouldn't reconsider the feature that lets the user edit album-level fields as well as item-level ones. I know that it's useful (and I originally suggested it!), but this may be a case where the benefit is outweighed by the extra complexity in the code (FakeAlbum and all that).

Yes, I definitely agree that things have gotten quite messy as a consequence of that change, and even if it is a highly desirable feature it is probably best to leave it out for the next iteration, sadly.

I'll work on reverting the changes related to editing both album- and item-level fields, plus using temporary ids, which should take care of most of the comments on the latest review!

sampsyo commented:
Awesome! I'm excited---and it will be interesting to tackle the album-level problem in the next iteration.

diego-plan9 commented on beetsplug/edit.py:
```diff
@@ -324,3 +333,55 @@ def save_changes(self, objs):
+        there are candidates) to the interactive importer prompt.
+        """
+        choices = [PromptChoice('d', 'eDit', self.importer_edit)]
+        if task.candidates:
```
This one actually revealed a problem if the user uses `Enter search` or `enter Id` to find a candidate not found during the initial lookup: as things stand, I believe this listener has no way of retrieving that candidate (as the new `candidate` it is handled *within* `TerminalImportSession.choose_match()`, and only saved when exiting the prompt by applying). I have added a note to the docs, but I'm wondering if something else should be done about it.

sampsyo replied:
Ah, good point. I also don't see an immediate solution, so let's revisit the problem later on.

diego-plan9 commented on beetsplug/edit.py:
```diff
@@ -313,8 +322,8 @@ def apply_data(self, objs, old_data, new_data):
-            id = int(old_dict['id'])
-            apply(obj_by_id[id], new_dict)
+            id_ = int(old_dict['id'])
+            apply_(obj_by_id[id_], new_dict)
```
These were renamed in order not to have overly-zealous IDEs complain about overriding a built-in symbol.

diego-plan9 commented:
Rolled back the two features that were the main cause of the increase in complexity (discarding completely the album-fields feature and adding a note to the documentation; and using temporary `id`s), and added some basic documentation and docstrings cleanup.

There are still some issues open which I hope did not get swallowed by github collapsing old diffs - the saving of the old data for comparison https://github.com/beetbox/beets/pull/1846#discussion_r52014937, and the merging of `RETAG` into `apply()` on the importer https://github.com/beetbox/beets/pull/1846#discussion_r52016245; but hopefully we are getting closer!

sampsyo commented:
Awesome! Thank you for the tremendous amount of work this took! :sparkles: This looks ready, IMO, so feel free to merge.

Woohoo!

diego-plan9 commented:
Cool, thanks for the feedback and for the help and guidance in bringing this iteration of the `edit` plugin to life! I still have a slight feeling of the feature falling a bit short of the potential user's expectations due to the (reasonable, on the other hand) compromises we had to make ... that will most likely result in another round of discussion and crazy ideas from me in regards to the plugin in the not-so-distant future! :repeat: 

Would it be ok to leave writing the changeset to you, as you'll probably be able to find a better way of summarizing the changes and conveying the *half-baked-ish* state of the feature?

Kraymer commented:
Kudos @diego-plan9 , impressive quality stuff ! :money_with_wings:

sampsyo commented:
Done! Let me know if I did the work justice, @diego-plan9.

diego-plan9 commented:
It does, thanks a lot for taking care of the changelog with such a neat choice of words! And thanks @Kraymer for the kind words, although I have to return the kudos for your fantastic amount of contributions to beets throughout the project's history!
